File: src/addressbook.ts
/**
 * @module addressbook
 * Address Book / Allowlist Guardrail.
 *
 * A configurable allowlist of known-good contract addresses the agent is
 * permitted to interact with. Any address NOT in the book is blocked.
 *
 * This is the simplest, most powerful defense: if you haven't explicitly
 * approved a contract, the agent cannot touch it.
 */

export type AddressBookResult = {
  decision: "ALLOW" | "BLOCK";
  reason: string;
  allowExecute: boolean;
};

export type AddressBookConfig = {
  /** Map of label → address (checksummed or lowercase, both accepted) */
  allowlist: Record<string, string>;
  /** If true, block all addresses not in the allowlist (default: true) */
  strict?: boolean;
};

/**
 * In-memory address allowlist.
 * Configure with known-good router addresses, token addresses, etc.
 */
export class AddressBook {
  private readonly addresses: Map<string, string>; // lowercase address → label
  private readonly strict: boolean;

  constructor(config: AddressBookConfig) {
    this.strict = config.strict ?? true;
    this.addresses = new Map();

    for (const [label, addr] of Object.entries(config.allowlist)) {
      if (!/^0x[a-fA-F0-9]{40}$/.test(addr)) {
        throw new Error(`invalid address for "${label}": ${addr}`);
      }
      this.addresses.set(addr.toLowerCase(), label);
    }
  }

  /**
   * Check if a target address is in the allowlist.
   */
  check(address: string): AddressBookResult {
    if (!/^0x[a-fA-F0-9]{40}$/.test(address)) {
      return {
        decision: "BLOCK",
        reason: `invalid address ${address} — BLOCK`,
        allowExecute: false,
      };
    }

    const normalized = address.toLowerCase();
    const label = this.addresses.get(normalized);

    if (label) {
      return {
        decision: "ALLOW",
        reason: `address is allowlisted as "${label}"`,
        allowExecute: true,
      };
    }

    if (this.strict) {
      return {
        decision: "BLOCK",
        reason: `UNKNOWN ADDRESS: ${address} is NOT in the allowlist. The agent cannot interact with unknown contracts. — BLOCK`,
        allowExecute: false,
      };
    }

    return {
      decision: "ALLOW",
      reason: `address not in allowlist but strict mode is off`,
      allowExecute: true,
    };
  }

  /** Check if an address is known */
  has(address: string): boolean {
    return this.addresses.has(address.toLowerCase());
  }

  /** Get the label for an address */
  labelOf(address: string): string | undefined {
    return this.addresses.get(address.toLowerCase());
  }

  /** Number of entries */
  size(): number {
    return this.addresses.size;
  }
}

File: src/addressbook.rs
//! Strict allowlist for contract targets.
//!
//! The simplest and strongest access control: unknown addresses BLOCK in
//! strict mode (the default you should use). Non-strict mode logs-but-allows
//! and is fail-open by configuration — only use it for discovery dry-runs.

use crate::types::GuardrailResult;
use std::collections::HashMap;

/// Lowercased-address → human label allowlist.
pub struct AddressBook {
    addresses: HashMap<String, String>, // lowercase address -> label
    strict: bool,
}

impl AddressBook {
    /// Build from `label → address` pairs. `Err` on any invalid address.
    /// `strict = true` is recommended (unknown → BLOCK). `strict = false`
    /// is fail-open by configuration — discovery dry-runs only; prefer
    /// [`AddressBook::new_strict`].
    pub fn new(allowlist: HashMap<String, String>, strict: bool) -> Result<Self, String> {
        let mut addresses = HashMap::new();
        for (label, addr) in allowlist {
            if !is_valid_eth_address(&addr) {
                return Err(format!("invalid address for \"{}\": {}", label, addr));
            }
            addresses.insert(addr.to_lowercase(), label);
        }
        Ok(Self { addresses, strict })
    }

    /// Strict constructor: unknown addresses BLOCK. Prefer this over
    /// `new(_, false)` everywhere except discovery dry-runs.
    pub fn new_strict(allowlist: HashMap<String, String>) -> Result<Self, String> {
        Self::new(allowlist, true)
    }

    /// Check `address`: invalid → BLOCK; known → ALLOW; unknown → BLOCK
    /// in strict mode, ALLOW (logged) otherwise.
    pub fn check(&self, address: &str) -> GuardrailResult {
        if !is_valid_eth_address(address) {
            return GuardrailResult::block(format!("invalid address {} — BLOCK", address));
        }

        let normalized = address.to_lowercase();
        if let Some(label) = self.addresses.get(&normalized) {
            GuardrailResult::allow(format!("address is allowlisted as \"{}\"", label))
        } else if self.strict {
            GuardrailResult::block(format!(
                "UNKNOWN ADDRESS: {} is NOT in the allowlist. The agent cannot interact with unknown contracts. — BLOCK",
                address
            ))
        } else {
            GuardrailResult::allow("address not in allowlist but strict mode is off")
        }
    }

    /// Membership test (case-insensitive).
    pub fn has(&self, address: &str) -> bool {
        self.addresses.contains_key(&address.to_lowercase())
    }

    /// Label for an allowlisted address, if present.
    pub fn label_of(&self, address: &str) -> Option<&String> {
        self.addresses.get(&address.to_lowercase())
    }

    /// Number of allowlisted entries.
    pub fn size(&self) -> usize {
        self.addresses.len()
    }
}

/// Strict `0x` + 40-hex validation. No EIP-55 checksum enforcement by
/// design (case is normalized on insert/lookup); checksum confusion is
/// out of scope — the allowlist comparison is case-insensitive.
pub fn is_valid_eth_address(addr: &str) -> bool {
    let clean = addr.trim();
    if !clean.starts_with("0x") && !clean.starts_with("0X") {
        return false;
    }
    let hex_part = &clean[2..];
    hex_part.len() == 40 && hex_part.chars().all(|c| c.is_ascii_hexdigit())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_addressbook_allowlisted() {
        let mut map = HashMap::new();
        map.insert(
            "USDC".to_string(),
            "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48".to_string(),
        );
        let book = AddressBook::new(map, true).unwrap();

        let res = book.check("0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48");
        assert!(res.allow_execute);
        assert!(res.reason.contains("allowlisted as \"USDC\""));
    }

    #[test]
    fn test_addressbook_strict_blocks_unknown() {
        let mut map = HashMap::new();
        map.insert(
            "USDC".to_string(),
            "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48".to_string(),
        );
        let book = AddressBook::new(map, true).unwrap();

        let res = book.check("0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045");
        assert!(!res.allow_execute);
        assert!(res.reason.contains("UNKNOWN ADDRESS"));
    }

    #[test]
    fn test_new_strict_blocks_unknown() {
        let mut map = HashMap::new();
        map.insert(
            "USDC".to_string(),
            "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48".to_string(),
        );
        let book = AddressBook::new_strict(map).unwrap();

        let res = book.check("0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045");
        assert!(!res.allow_execute);
    }
}
